### ISSUE TITLE
Converting a positive subnormal double to float yields +infinity

### ISSUE THREAD

**Issue by andreast271** [diffblue/cbmc#7616]
CBMC version: 5.79.0
Operating system: Linux x86_64
Exact command line resulting in the issue: cbmc main.c --trace
What behaviour did you expect: verification successful
What happened instead: verification failed with an incorrect trace

The example program is
```
int main() {
  float f;
  double d;
  if(d > 0.0 && d < 1.0) {
    f = d;
    __CPROVER_assert(f <= 1.0f, "limited");
  }
  return 0;
}
```
CBMC reports a failure for the assertion. In the trace (see attached trace.log), d is shown as a positive subnormal number and f is shown as positive infinity:
```
State 12 file d2f.c function main line 3 thread 0
----------------------------------------------------
  d=5.562684e-309 (00000000 00000011 11111111 11111111 11111110 00000100 00000000 00000000)

State 14 file d2f.c function main line 5 thread 0
----------------------------------------------------
  f=+INFINITY (01111111 10000000 00000000 00000000)
```

If d is restricted to normal numbers, the assertion passes.
Not every subnormal number converts to +infinity.
If the assertion is changed to state that f is in the expected range or positive infinity, it passes:
```
 __CPROVER_assert((f >= 0 && f <= 1.0f) || (isinf(f) && f > 0), "limited"); // this passes
```

[main.c.txt](https://github.com/diffblue/cbmc/files/11059827/main.c.txt)
[trace.log](https://github.com/diffblue/cbmc/files/11059840/trace.log)

**Comment by tautschnig** [comment]
It seems the problem is limited to our flattening back-end. The following instantiation of the counterexample passes:
```
#include <assert.h>
#include <stdint.h>
#include <stdio.h>

union U
{
  double d;
  uint64_t b;
};

int main() {
  union U u = { .b = 0b0000000000000011111111111111111111111110000001000000000000000000 };
  float f;
  double d = u.d;
  if(d > 0.0 && d < 1.0) {
    f = d;
    printf("%x\n", f);
    assert(f <= 1.0f);
  }
  return 0;
}
```
But continues to fail when running `cbmc --no-simplify`.

**Comment by martin-cs** [comment]
This is definitely a bug.  Definitely with the encoding.  It is probably an underflow somewhere in the normalise / denormalise step of rounding.  I thought I had fixed all of these back in 2014?  I'll have another look.

**Comment by tautschnig** [comment]
#7617 makes a first attempt at a fix, but it still seems to be incomplete (works fine for the original counterexample, but doesn't seem to produce the right result for a different counterexample).

**Comment by martin-cs** [comment]
So the issue is likely here:
https://github.com/diffblue/cbmc/blob/9f894aeb9371817146cb2ec229785baee566b3e8/src/solvers/floatbv/float_utils.cpp#L215

`unpack` should work but my guess is that the rounder is not handling the unpacked subnormals.  In principle the design is that the rounder should take any unpacked float and round it to the given spec.  There are a few issues with this.  Not knowing the input format means that subnormal inputs can be missed.  Also the space of unpacked floats / target specs is not fully used in normal operation so it is possible for there to be subtle bugs near the edges of use.

Again, my /guess/ is that https://github.com/diffblue/cbmc/blob/9f894aeb9371817146cb2ec229785baee566b3e8/src/solvers/floatbv/float_utils.cpp#L921 is roughly the location of the bug and there will be an overflow or an underflow in the exponent caused by not extending it enough when normalising.  This will wind up with the exponent being too high and then `round_exponent` will set it to infinity.

https://github.com/diffblue/cbmc/blob/9f894aeb9371817146cb2ec229785baee566b3e8/src/solvers/floatbv/float_utils.cpp#L825
is where it would over/underflow.

https://github.com/diffblue/cbmc/blob/9f894aeb9371817146cb2ec229785baee566b3e8/src/solvers/floatbv/float_utils.cpp#L835
note that this is the bias for the target format; in this case single, *not* the bias in the source format.

**Comment by tautschnig** [comment]
Thanks to @martin-cs's guidance I believe I have managed to come up with the fix: #7617.
